### PR TITLE
Update SDP docs

### PR DESCRIPTION
**Pull request by erastorgueva-nv**
# What does this PR do ?

Add more detail to SDP section in NeMo docs, and add link to that page in the NeMo README.rst.

**Collection**: docs

# Changelog 
- update SDP section in NeMo docs
- update NeMo README

# Usage
N/A

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [N/A] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ N/A ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [x] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: docs/source/tools/speech_data_processor.rst [hunk 1]
```diff
@@ -5,4 +5,134 @@ Speech Data Processor (SDP) is a toolkit to make it easy to:
...
+
+Overview of how SDP processes a dataset
+---------------------------------------
+
+1. You call the ``main.py`` script, passing in a YAML config file, possibly with some overrides.
+2. ``main.py`` script calls ``run_processors.py``, passing in your config.
+3. ``run_processors.py`` does the following:
+
+a. picks out the processors you wish to run (you can specify a subset of the processors in the config override, e.g. to avoid re-running time-consuming steps).
+b. if some of the processors have not had "output_manifest_file" or "input_manfiest_file" entries specified, SDP will automatically create temporary files for those.
```
Comment (titu1994) [inline docs/source/tools/speech_data_processor.rst:22]: Perhaps some explanation of what these args mean in a manifest before detailing they will be created automatically.

File: docs/source/tools/speech_data_processor.rst [hunk 2]
```diff
@@ -5,4 +5,134 @@ Speech Data Processor (SDP) is a toolkit to make it easy to:
...
+c. instantiate the processor classes using ``hydra.utils.instantiate``
+d. run the run-time processor tests by calling the ``processor.test()`` method.
```
Comment (titu1994) [inline docs/source/tools/speech_data_processor.rst:24]: Run a test during actual processing ?

File: docs/source/tools/speech_data_processor.rst [hunk 3]
```diff
@@ -5,4 +5,134 @@ Speech Data Processor (SDP) is a toolkit to make it easy to:
...
+e. run the processing method (``processor.process()``) of each processor in order.
+  
+
+Layout of config YAML files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The YAML config file for processing a dataset must contain a key ``processors``, the value of which is a list. Each item in that list is expected to be a dictionary specifying a processor class, i.e. it must have a key ``_target_``, the value of which is a path to a "processor" class, and the remaining keys must be the kwargs necessary to instantiate that class with ``hydra.utils.instantiate()`` (c.f. https://hydra.cc/docs/advanced/instantiate_objects/overview/).
+
+SDP will run the processors specified in the ``processors`` list in the config file. It will also check for a ``processors_to_run`` key in the config file, which can be either the string "all", or any Python "slice" object.
```
Comment (titu1994) [inline docs/source/tools/speech_data_processor.rst:33]: Maybe have an example of a call here (or after the note)

File: docs/source/tools/speech_data_processor.rst [hunk 4]
```diff
@@ -5,4 +5,134 @@ Speech Data Processor (SDP) is a toolkit to make it easy to:
...
+
+Processor classes
+-----------------
+
+
+**BaseProcessor**
+~~~~~~~~~~~~~~~~~~
+
+All processor classes inherit from the ``BaseProcessor`` class. This is a simple abstract class which has 2 empty methods: ``process()`` and ``test()``. 
+These serve to remind us that SDP essentially just runs ``test()`` on all processors, and then ``process()`` on all processors.
```
Comment (titu1994) [inline docs/source/tools/speech_data_processor.rst:46]: Why is it called test ()? Ie what is it testing ?

File: docs/source/tools/speech_data_processor.rst [hunk 5]
```diff
@@ -5,4 +5,134 @@ Speech Data Processor (SDP) is a toolkit to make it easy to:
...
+
+``ASRInference`` is a child class of ``BaseProcessor``. It has a simple ``process()`` method which runs transcription on every utterance in the input_manifest.
```
Comment (titu1994) [inline docs/source/tools/speech_data_processor.rst:48]: What is it's test() method ? Is it important enough to note?

File: docs/source/tools/speech_data_processor.rst [hunk 6]
```diff
@@ -5,4 +5,134 @@ Speech Data Processor (SDP) is a toolkit to make it easy to:
...
+
+``WriteManifest`` is also a child class of ``BaseProcessor``. It has a simple ``process()`` method which saves a copy of the input manifest containing only the fields specified in ``fields_to_save``.
+
+**BaseParallelProcessor**
+~~~~~~~~~~~~~~~~~~~~~~~~~
+``BaseParallelProcessor`` inherits from the ``BaseProcessor`` class. Within its ``.process()`` method, it calls other methods and functions, which allow it to do more complex processing. 
+Most importantly, it calls its ``.process_dataset_entry(data_entry)`` method on every utterance in the manifest, and it does this in parallel, allowing for more efficient processing.
```
Comment (titu1994) [inline docs/source/tools/speech_data_processor.rst:55]: . Before name is mistake or referring to method in some special way?

File: docs/source/tools/speech_data_processor.rst [hunk 7]
```diff
@@ -5,4 +5,134 @@ Speech Data Processor (SDP) is a toolkit to make it easy to:
...
+* ``test_cases`` (optional, list of dicts) - test cases for checking that the processor makes the changes that we are expecting.
+
+``ModifyManifestTextProcessor`` has the following methods: 
+* ``.test()``: this method makes sure that the output from the processor matches the expected output specified in the ``test_cases`` parameter.
+* ``.process_dataset_entry(data_entry)``: this method applies processing to a ``data_entry``. First, spaces are added to the start and end of the 'text' and 'pred_text' entries (if they exist), then the abstract method ``._process_dataset_entry(data_entry)`` is called. Then, any extra spaces (e.g. two spaces next to each other '  ') are removed from 'text' and 'pred_text' entries.
+* ``._process_dataset_entry(data_entry)``: this is an abstract method which will be over-written by children of ``ModifyManifestTextProcessor``.
```
Comment (titu1994) [inline docs/source/tools/speech_data_processor.rst:94]: Hmm I see now what the . Is. Let's replace it with full base class name everywhere

**Review by titu1994 — COMMENTED** [review]
Overall looks fantastic. Highly detailed and complete info, just needs a bit of reordering to make a few things clear

File: docs/source/tools/speech_data_processor.rst (hunk 6 above)
Comment (erastorgueva-nv) [reply docs/source/tools/speech_data_processor.rst:55]: Mistake, nothing special

File: docs/source/tools/speech_data_processor.rst (hunk 7 above)
Comment (erastorgueva-nv) [reply docs/source/tools/speech_data_processor.rst:94]: OK, I will put full base class name in some places, in others just remove the "."

**Review by titu1994 — APPROVED** [review]
Excellent docs! Side note - maybe a mini tutorial following the instructions here of using and creating custom processors on some toy AN4 data would be a nice showcase of SDP but it's not a requirement. The docs are clear and precise so it should be easy to follow

**Review by titu1994 — DISMISSED** [review]
Excellent docs! Side note - maybe a mini tutorial following the instructions here of using and creating custom processors on some toy AN4 data would be a nice showcase of SDP but it's not a requirement. The docs are clear and precise so it should be easy to follow

**Pull request merged**